### ISSUE TITLE
[Bug]: panic: runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by svenstaro** [hetznercloud/terraform-provider-hcloud#491]
### What happened?

Running the provider it will _sometimes_ (definitely not always) result in this stack trace:
```
Releasing state lock. This may take a few moments...
Stack trace from the terraform-provider-hcloud_v1.27.0 plugin:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xaf8c56]
goroutine 63 [running]:
github.com/hetznercloud/hcloud-go/hcloud.(*ActionClient).WatchOverallProgress.func1(0xc0000c6a80, 0xc000724840, 0xc0008281d0, 0x1, 0x1, 0xc000478178, 0x10fea38, 0xc000716fc0)
	github.com/hetznercloud/hcloud-go@v1.27.0/hcloud/action.go:196 +0x176
created by github.com/hetznercloud/hcloud-go/hcloud.(*ActionClient).WatchOverallProgress
	github.com/hetznercloud/hcloud-go@v1.27.0/hcloud/action.go:189 +0xde
Error: The terraform-provider-hcloud_v1.27.0 plugin crashed!
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```
After upgrading to 1.32.2 I still get this:
```
Releasing state lock. This may take a few moments...
Stack trace from the terraform-provider-hcloud_v1.32.2 plugin:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa6b40f]
goroutine 130 [running]:
github.com/hetznercloud/hcloud-go/hcloud.(*ActionClient).WatchOverallProgress.func1()
	github.com/hetznercloud/hcloud-go@v1.33.1/hcloud/action.go:196 +0x1af
created by github.com/hetznercloud/hcloud-go/hcloud.(*ActionClient).WatchOverallProgress
	github.com/hetznercloud/hcloud-go@v1.33.1/hcloud/action.go:189 +0x133
Error: The terraform-provider-hcloud_v1.32.2 plugin crashed!
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```
Which looks exactly like the same stacktrace so we can see that this has been in here for some time. I'm not sure this is a problem in this provider or in hcloud-go but I thought starting here from the provider would be fair.

The lines leading up to the stacktrace are:
```
Acquiring state lock. This may take a few moments...
module.dynamic_environments["test"].hcloud_volume_attachment.database: Destroying... [id=1337]
module.dynamic_environments["test"].hcloud_server_network.database: Destroying... [id=1337-1337]
module.dynamic_environments["test"].hcloud_server_network.database: Still destroying... [id=1337-1337, 10s elapsed]
module.dynamic_environments["test"].hcloud_volume_attachment.database: Still destroying... [id=1337, 10s elapsed]
╷
│ Warning: Applied changes may be incomplete
│ 
│ The plan was created with the -target option in effect, so some changes
│ requested in the configuration may have been ignored and the output values
│ may not be fully updated. Run the following command to verify that no other
│ changes are pending:
│     terraform plan
│ 	
│ Note that the -target option is not suitable for routine use, and is
│ provided only for exceptional situations such as recovering from errors or
│ mistakes, or when Terraform specifically suggests to use it as part of an
│ error message.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
```

### What did you expect to happen?

It shouldn't crash.

### Please provide a minimal working example

It's really hard to pinpoint the exact cause as it's not perfectly reproducible but it seems to always occur when destroying a related `hcloud_volume_attachment`, `hcloud_server_network` both in the same destruction run.

**Comment by apricote** [comment]
We had multiple issues opened with the same stacktrace. As this issue contains the most information, I will close all others and redirect any questions here.

---

This issue is hard to reproduce, it would be great if someone could provide us with debug logs for a `terraform apply` run where this happened.

To get debug logs, you need to run with `TF_LOG=DEBUG TF_LOG_PATH=watch-action-nil-pointer.log terraform apply`.

You can then send the debug logs to my email: `julian.toelle <at> hetzner-cloud.de`.

**Comment by sasoiliev** [comment]
> You can then send the debug logs to my email: julian.toelle <at> hetzner-cloud.de.

@apricote, I just sent a debug log from a run that reproduced the issue to your email as per your request.

**Comment by apricote** [comment]
Thanks a lot @sasoiliev! With the logs I was able to figure out where the problem was coming from. Relevant excerpt from the logs:

```
2022-12-21T12:29:31.446+0200 [INFO]  provider.terraform-provider-hcloud_v1.35.2: --- Request:
POST /v1/servers/REDACTED/actions/detach_from_network HTTP/1.1
Host: api.hetzner.cloud
User-Agent: hcloud-terraform/not build yet hcloud-go/1.35.2
Content-Length: 19
Authorization: REDACTED
Content-Type: application/json
Accept-Encoding: gzip

{"network":REDACTED}: timestamp=2022-12-21T12:29:31.446+0200
2022-12-21T12:29:32.323+0200 [INFO]  provider.terraform-provider-hcloud_v1.35.2: --- Response:
HTTP/2.0 500 Internal Server Error
Content-Length: 90
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Content-Type: application/json
Date: Wed, 21 Dec 2022 10:29:32 GMT
Ratelimit-Limit: 3600
Ratelimit-Remaining: REDACTED
Ratelimit-Reset: REDACTED
Strict-Transport-Security: max-age=15724800; includeSubDomains
X-Correlation-Id: REDACTED

{
  "error": {
    "message": null,
    "code": "server_error",
    "details": null
  }
}: timestamp=2022-12-21T12:29:32.323+0200
2022-12-21T12:29:32.325+0200 [DEBUG] provider.terraform-provider-hcloud_v1.35.2: panic: runtime error: invalid memory address or nil pointer dereference
2022-12-21T12:29:32.325+0200 [DEBUG] provider.terraform-provider-hcloud_v1.35.2: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xaedc97]
2022-12-21T12:29:32.325+0200 [DEBUG] provider.terraform-provider-hcloud_v1.35.2: 
2022-12-21T12:29:32.325+0200 [DEBUG] provider.terraform-provider-hcloud_v1.35.2: goroutine 472 [running]:
2022-12-21T12:29:32.325+0200 [DEBUG] provider.terraform-provider-hcloud_v1.35.2: github.com/hetznercloud/hcloud-go/hcloud.(*ActionClient).WatchOverallProgress.func1()
2022-12-21T12:29:32.325+0200 [DEBUG] provider.terraform-provider-hcloud_v1.35.2: 	github.com/hetznercloud/hcloud-go@v1.35.2/hcloud/action.go:196 +0x1b7
2022-12-21T12:29:32.325+0200 [DEBUG] provider.terraform-provider-hcloud_v1.35.2: created by github.com/hetznercloud/hcloud-go/hcloud.(*ActionClient).WatchOverallProgress
2022-12-21T12:29:32.325+0200 [DEBUG] provider.terraform-provider-hcloud_v1.35.2: 	github.com/hetznercloud/hcloud-go@v1.35.2/hcloud/action.go:189 +0x12e
2022-12-21T12:29:32.326+0200 [DEBUG] provider: plugin process exited: path=/tmp/tmp95tdg52d/terraform/providers/registry.terraform.io/hetznercloud/hcloud/1.35.2/linux_amd64/terraform-provider-hcloud_v1.35.2 pid=4143441 error="exit status 2"
```

The call to `client.Server.DetachFromNetwork` failed, but we did not properly handle the error before passing the empty `Action` pointer to `hcclient.WaitForAction`. This will be fixed in #602.
